### PR TITLE
feat(http): Add authentication support

### PR DESCRIPTION
**Pull request by royjacobson**
We should probably protect the metrics page somehow, this is the quickest way to do it.

This is currently untested because testing is kind of a hassle (no http unit tests at all currently). However I tested it works locally and I'm willing to write a Python test with `requests` once it's merged :)

### CODE REVIEW

**Comment by romange** [comment]
I am not sure how prometheus will access it.
Moreover, not sure how humans can access it.

**Comment by royjacobson** [comment]
It's pretty standard. The browser prompt looks like this:
image

I think all the prometheus tooling will support it, but cc @ashotland

File: util/http/http_handler.cc [hunk 1]
```diff
@@ -366,9 +367,44 @@ void HttpConnection::HandleRequests() {
...
+    return false;
+  }
+
+  return true;
+}
+
 void HttpConnection::HandleSingleRequest(const RequestType& req, HttpContext* cntx) {
   CHECK(owner_);
 
+  if (!CheckRequestAuthorization(req, cntx))
```
Comment (romange) [inline util/http/http_handler.cc:405]: lets at least do it only if password is not empty

Comment (royjacobson) [reply util/http/http_handler.cc]: done

File: util/http/http_handler.h [hunk 2]
```diff
@@ -80,6 +80,12 @@ class HttpListenerBase : public ListenerInterface {
     enable_metrics_ = true;
   }
 
+  // Enforce authentication to this HTTP server with the user "user"
+  // and this password.
+  void SetPassword(std::string password) {
```
Comment (chakaz) [inline util/http/http_handler.h:85]: Correct me if I'm wrong, but this is dead code, right? (i.e. nothing calls `SetPassword()`)
Where do you intend to call this method?

File: util/http/http_handler.cc [hunk 3]
```diff
@@ -366,9 +367,44 @@ void HttpConnection::HandleRequests() {
   LOG_IF(INFO, !FiberSocketBase::IsConnClosed(ec)) << "Http error " << ec.message();
 }
 
+bool HttpConnection::CheckRequestAuthorization(const RequestType& req, HttpContext* cntx) {
+  CHECK(owner_);
+
+  bool authenticated = false;
+  if (const boost::string_view header = req[h2::field::authorization]; !header.empty()) {
+    if (header.substr(0, 6) == "Basic ") {
```
Comment (chakaz) [inline util/http/http_handler.cc:375]: Use `absl::StartsWith()`?

File: util/http/http_handler.h (hunk 2 above)
Comment (royjacobson) [reply util/http/http_handler.h:85]: From `facade::Connection::HandleRequests`.

Comment (chakaz) [reply util/http/http_handler.h:85]: Ok, please add `override` then :)

Comment (royjacobson) [reply util/http/http_handler.h:85]: But I now added a demonstration to `http_main`

Comment (royjacobson) [reply util/http/http_handler.h:85]: It's not virtual though?

Comment (chakaz) [reply util/http/http_handler.h:85]: Now I understand, thanks!

**Review by chakaz — APPROVED** [review]

**Pull request merged**